cfg/CFG.cc

```diff
@@ -22,97 +32,161 @@ BasicBlock *CFG::freshBlock(int outerLoops, int rubyBlockId) {
     return r.get();
 }
 
+void CFG::enterLocalInternal(core::LocalVariable variable, LocalRef &ref) {
+    ENFORCE_NO_TIMER(!this->localVariablesFrozen);
+    int id = this->maxVariableId++;
+    this->localVariables.emplace_back(variable);
+
+    // Default values
+    this->minLoops.emplace_back(INT_MAX);
+    this->maxLoopWrite.emplace_back(0);
+
+    ENFORCE(this->localVariables.size() == this->minLoops.size());
+    ENFORCE(this->localVariables.size() == this->maxLoopWrite.size());
+    ref = LocalRef(id);
+}
+
+LocalRef CFG::enterLocal(core::LocalVariable variable) {
+    auto &ref = localVariableToLocalRef[variable];
+    if (!ref.exists() && variable.exists()) {
+        // ref is an out parameter.
+        enterLocalInternal(variable, ref);
+    }
+    return ref;
+}
+
 CFG::CFG() {
     freshBlock(0, 0); // entry;
     freshBlock(0, 0); // dead code;
     deadBlock()->bexit.elseb = deadBlock();
     deadBlock()->bexit.thenb = deadBlock();
-    deadBlock()->bexit.cond.variable = core::LocalVariable::unconditional();
+    deadBlock()->bexit.cond.variable = LocalRef::unconditional();
+
+    UnfreezeCFGLocalVariables unfreezeVars(*this);
+    // Enter a few fixed local variables
+    // noVariable is special because it doesn't 'exist'.
+    this->enterLocalInternal(core::LocalVariable::noVariable(),
+                             localVariableToLocalRef[core::LocalVariable::noVariable()]);
+
+    LocalRef blockCall = this->enterLocal(core::LocalVariable::blockCall());
+    ENFORCE(blockCall == LocalRef::blockCall());
+
+    LocalRef selfVariable = this->enterLocal(core::LocalVariable::selfVariable());
+    ENFORCE(selfVariable == LocalRef::selfVariable());
+
+    LocalRef unconditional = this->enterLocal(core::LocalVariable::unconditional());
+    ENFORCE(unconditional == LocalRef::unconditional());
+
+    LocalRef finalReturn = this->enterLocal(core::LocalVariable(core::Names::finalReturn(), 0));
+    ENFORCE(finalReturn == LocalRef::finalReturn());
 }
 
 CFG::ReadsAndWrites CFG::findAllReadsAndWrites(core::Context ctx) {
     Timer timeit(ctx.state.tracer(), "findAllReadsAndWrites");
     CFG::ReadsAndWrites target;
+    target.readsSet.resize(maxBasicBlockId);
     target.reads.resize(maxBasicBlockId);
     target.writes.resize(maxBasicBlockId);
     target.dead.resize(maxBasicBlockId);
-    vector<UnorderedSet<core::LocalVariable>> readsAndWrites(maxBasicBlockId);
```

Comment:
This is great, I really like not having to keep track of this during the main loop!

Reply:
Ditto.